barretenberg/cpp/src/barretenberg/vm/avm/trace/gadgets/range_check.hpp

```diff
@@ -0,0 +1,133 @@
+#pragma once
+
+#include "barretenberg/vm/avm/generated/relations/range_check.hpp"
+#include "barretenberg/vm/avm/trace/common.hpp"
+#include <cstdint>
+
+enum class EventEmitter {
+    ALU,
+    MEMORY,
+};
+
+namespace bb::avm_trace {
+// template <class TraceUtilisingRange>
+class AvmRangeCheckBuilder {
+  public:
+    struct RangeCheckEvent {
+        uint32_t clk;
+        FF value;
+        uint8_t num_bits;
+        EventEmitter emitter;
+    };
+
+    struct RangeCheckEntry {
+        uint32_t clk;
+        FF value;
+        uint8_t num_bits;
+        // 8 total 16-bit registers, the last one is dynamic
+        std::array<uint16_t, 7> fixed_slice_registers;
+        uint16_t dynamic_slice_register;
+        // The number of bits that need to be dynamically checked
+        uint16_t dyn_bits;
+        // The difference between
+        uint16_t dyn_diff;
+        // Bit string representing which of the is_lte_x flags are on
+        // From LSB to MSB:
+        // [is_lte_u16, is_lte_u32, is_lte_u48, is_lte_u64, is_lte_u80, is_lte_u96, is_lte_u112, is_lte_u128]
+        uint8_t bit_range_flag;
+    };
+
+    std::array<std::unordered_map<uint16_t, uint32_t>, 8> u16_range_chk_counters;
+    std::unordered_map<uint16_t, uint32_t> powers_of_2_counts;
+    std::unordered_map<uint16_t, uint32_t> dyn_diff_counts;
+
+    // This function just enqueues a range check event, we handle processing them later in finalize.
+    bool assert_range(FF value, uint8_t num_bits, EventEmitter e, uint32_t clk)
+    {
+        // We don't support range checks on values that are field-sized
+        // ASSERT(num_bits <= 128);
+        range_check_events.push_back({ clk, value, num_bits, e });
+        return true;
+    }
+
+    // Turns range check events into real entries
+    std::vector<RangeCheckEntry> finalize()
+    {
+        std::vector<RangeCheckEntry> entries;
+        // Process each range check event into entries
+        for (auto& event : range_check_events) {
+            auto entry = RangeCheckEntry{};
+            // Set all the easy stuff
+            entry.clk = event.clk;
+            entry.value = event.value;
+            entry.num_bits = event.num_bits;
+            auto value_u256 = uint256_t(event.value);
+
+            // Now some harder stuff, split the value into 16-bit chunks
+            for (size_t i = 0; i < 8; i++) {
+                // The most significant 16-bits have to be placed in the dynamic slice register
+                if (event.num_bits <= 16) {
+                    entry.dynamic_slice_register = uint16_t(value_u256);
+                    u16_range_chk_counters[7][entry.dynamic_slice_register]++;
```

Comment:
We discussed about separating finalize and counters, it might be needed.

Also, consider using the "new" pattern I'm establishing of having finalize take the main trace and return nothing (I had to do it because some trace was using shifts -- gas -- and I needed to set something on the current and next row).